working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -76,36 +76,66 @@ Searching our project, we find 4 dependencies:
 
-Examining the code created by Copilot, the scenarios identified are great; we have tests for the nominal case and a couple edge cases of invalid or non-existent book ids. However, there are a number of issues:
-- imports were added in the middle of the file, one of which is tacked on to the last line of an existing test
-- `test_validate_book_with_valid_id` doesn't use our `two_saved_books` test fixture to ensure that there are books in the database before the test runs
-- `test_validate_book_with_valid_id` thinks the result of `validate_book(book_id)` will be an integer rather than a `Book` instance
-- `test_validate_book_with_invalid_id` and `test_validate_book_with_nonexistent_id` try to read the error value from outside the `with pytest.raises` block where it's out of scope
+> Please rewrite the validate_book tests to import and directly invoke the validate_book function
+
+Copilot's response is looking better, in our case Copilot added the import for `validate_book` at the top of the file and generated the following tests:
+```py
+from app.routes import validate_book
+
+...
+
+def test_validate_book_succeeds(client, two_saved_books):
+    # Act
+    book = validate_book(1)
+
+    # Assert
+    assert book.id == 1
+    assert book.title == "Ocean Book"
+    assert book.description == "watr 4evr"
+
+def test_validate_book_missing_record(client, two_saved_books):
+    # Act & Assert
+    with pytest.raises(NotFound):
+        validate_book(3)
+
+def test_validate_book_invalid_id(client, two_saved_books):
+    # Act & Assert
+    with pytest.raises(BadRequest):
+        validate_book("cat")
+```
+
+These tests are getting closer to what we need, but still won't work for us. The main issues are that:
+- The `validate_book` import path at the top of the file is slightly wrong
+- The tests require `pytest` to use `with pytest.raises`
+- The symbols `NotFound` and `BadRequest` used in the edge case tests do not exist
+- The status code isn't being checked on the edge cases where the request should have been aborted
+- The `client` fixture is not required on any of the tests since we aren't making a request to an endpoint
+
+Fixing the imports is a couple lines of code, but for our edge cases, we may need to do some research to be sure of the error we expect to be raised by our functions and ensure that our tests are looking for that particular error to be raised. In this case, we know from the Building an API series that the `validate_book` function will raise an `HTTPException` that we need to import from `werkzeug.exceptions` to access in our tests.
 
 ### !callout-info
```

Comment:
If only we could make this callout appear flashing in 1000pt font.

With the model that's currently in use, it seems like there's a wider variation in the results we're getting (mine has jumped steps in a few places compared to the text), so we'll really need to set folks expectations to read this as one possible flow through the experience, and that there's is likely to differ significantly.

Reply:
I can add a reminder at the outset of each lesson that we are working with a generative tool and what they see while following the same steps may vary greatly from the results in the example we are walking through in the lesson.